src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -17,15 +17,40 @@ open VectorMatrix
 module Mpqf = SharedFunctions.Mpqf
 
 module Rhs = struct
-  (* (Some i, k) represents a sum of a variable with index i and the number k.
-     (None, k) represents the number k. *)
-  type t = (int option * GobZ.t) [@@deriving eq, ord, hash]
-  let var_zero i = (Some i, Z.zero)
-  let show_formatted formatter = function
-    | (Some v, o) when Z.equal o Z.zero -> formatter v
-    | (Some v, o) -> Printf.sprintf "%s%+Ld" (formatter v) (Z.to_int64 o)
-    | (None, o) -> Printf.sprintf "%Ld" (Z.to_int64 o)
-  let show rhs = show_formatted (Printf.sprintf "var_%d") rhs
+  (* Rhs represents coefficient*var_i + offset / divisor
+     depending on whether coefficient is 0, the monomial term may disappear completely, not refering to any var_i, thus:
+     (Some (coefficient, i), offset, divisor )         with coefficient != 0    , or
+     (None                 , offset, divisor ) *)
+  type t = ((GobZ.t * int) option *  GobZ.t * GobZ.t) [@@deriving eq, ord, hash]
+  let var_zero i = (Some (Z.one,i), Z.zero, Z.one)
+  let show_coeff c =
+    if Z.equal c Z.one then ""
+    else if Z.equal c Z.minus_one then "-"
+    else (Z.to_string c) ^"·"
+  let show_rhs_formatted formatter = function
+    | (Some (coeff,v), o,_) when Z.equal o Z.zero -> Printf.sprintf "%s%s" (show_coeff coeff) (formatter v)
+    | (Some (coeff,v), o,_) -> Printf.sprintf "%s%s%+Ld" (show_coeff coeff) (formatter v) (Z.to_int64 o)
+    | (None,   o,_) -> Printf.sprintf "%Ld" (Z.to_int64 o)
+  let show (v,o,d) =
+    let rhs=show_rhs_formatted (Printf.sprintf "var_%d") (v,o,d) in
+    if not (Z.equal d Z.one) then "(" ^ rhs ^ ")/" ^ (Z.to_string d) else rhs
+
+  (** factor out gcd from all terms, i.e. ax=by+c is the canonical form for adx+bdy+cd *)
+  let canonicalize (v,o,d) =
+    let gcd = Z.gcd o d in
+    let gcd = match v with
+      | Some (c,_) -> Z.gcd c gcd
+      | None -> gcd
+    in let gcd = if (Z.(lt d  Z.zero)) then Z.neg gcd else gcd in
```

Comment:
```suggestion
    in
    let gcd = if Z.(lt d zero) then Z.neg gcd else gcd in
```

Also: Why is it only a problem if `d` is negative, what if the other components are also negative and make the gcd negative?

Reply:
This is part of the precondition; d is supposed to be positive; in order to make it positive, we use gcd as a carrier vessel to consistently flip signs

Reply:
If I understood the code correctly `d` is the divider i.e. the `a` in the formular of the comment, `v` is the variable `y` and its coefficient `b` and `o` is the offset `c`. I guess `d` as divider should be positive such that `ax` is positive in the end. Did I understand it correctly?